lib/tests/complex.rs

```diff
@@ -0,0 +1,102 @@
+mod parse;
+
+use surrealdb::sql::Value;
+use surrealdb::Datastore;
+use surrealdb::Error;
+use surrealdb::Session;
+
+async fn run_queries(
+	sql: &str,
+) -> Result<impl Iterator<Item = Result<Value, Error>> + ExactSizeIterator + 'static, Error> {
+	let dbs = Datastore::new("memory").await?;
+	let ses = Session::for_kv().with_ns("test").with_db("test");
+	dbs.execute(&sql, &ses, None, false).await.map(|v| v.into_iter().map(|res| res.result))
+}
+
+#[tokio::test]
+async fn self_referential_field() -> Result<(), Error> {
+	let mut res = run_queries(
+		"
+		CREATE pet:dog SET tail = <future> { tail };
+	",
+	)
+	.await?;
+
+	assert_eq!(res.len(), 1);
+
+	let result = res.next().unwrap();
+	assert!(matches!(result, Err(Error::ComputationDepthExceeded)), "got: {:?}", result);
+
+	Ok(())
+}
+
+#[tokio::test]
+async fn cyclic_fields() -> Result<(), Error> {
+	let mut res = run_queries(
+		"
+		CREATE recycle SET consume = <future> { produce }, produce = <future> { consume };
+	",
+	)
+	.await?;
+
+	assert_eq!(res.len(), 1);
+
+	let result = res.next().unwrap();
+	assert!(matches!(result, Err(Error::ComputationDepthExceeded)), "got: {:?}", result);
+
+	Ok(())
+}
+
+#[tokio::test]
+#[cfg(not(debug_assertions))]
```

Comment:
Two of the tests unfortunately don't work in `debug` mode because the stack is smaller and the unoptimized functions take more of it.

Reply:
Fixed this by spawning a thread with more stack while in debug mode.